From a633ef2d7871a0d29b941c0ef17efd32d5689a93 Mon Sep 17 00:00:00 2001
Message-Id: <a633ef2d7871a0d29b941c0ef17efd32d5689a93.1398812351.git.mark.gross@intel.com>
In-Reply-To: <e87e3533536ae5eec0e06ceb7ef7428e42855790.1398812350.git.mark.gross@intel.com>
References: <e87e3533536ae5eec0e06ceb7ef7428e42855790.1398812350.git.mark.gross@intel.com>
From: Lejun Zhu <lejun.zhu@intel.com>
Date: Thu, 27 Feb 2014 01:16:34 +0800
Subject: [PATCH 206/429] sdhci-acpi: use gpio pin for card detection

Current sdhci-acpi driver doesn't detect the presence of SD card
on Asus T100. This patch calls mmc_gpio_request_cd() so that
the card detection routine will use the GPIO pin status instead of
the Present State register.

This patch requires the GPIO patch to work. Verified on T100 and
Dell v11p.

Change-Id: I18ad6a5a107cbffa0256959e1cca4ae28f70b82c
---
 drivers/mmc/host/sdhci-acpi.c | 28 ++++++----------------------
 1 file changed, 6 insertions(+), 22 deletions(-)

diff --git a/drivers/mmc/host/sdhci-acpi.c b/drivers/mmc/host/sdhci-acpi.c
index 9ce17f6..faeb7a3 100644
--- a/drivers/mmc/host/sdhci-acpi.c
+++ b/drivers/mmc/host/sdhci-acpi.c
@@ -41,6 +41,7 @@
 #include <linux/mmc/host.h>
 #include <linux/mmc/pm.h>
 #include <linux/mmc/sdhci.h>
+#include <linux/mmc/slot-gpio.h>
 
 #include "sdhci.h"
 
@@ -194,17 +195,10 @@ static const struct sdhci_acpi_slot *sdhci_acpi_get_slot(acpi_handle handle,
 
 #ifdef CONFIG_PM_RUNTIME
 
-static irqreturn_t sdhci_acpi_sd_cd(int irq, void *dev_id)
-{
-	mmc_detect_change(dev_id, msecs_to_jiffies(200));
-	return IRQ_HANDLED;
-}
-
 static int sdhci_acpi_add_own_cd(struct device *dev, struct mmc_host *mmc)
 {
 	struct gpio_desc *desc;
-	unsigned long flags;
-	int err, irq;
+	int err, gpio;
 
 	desc = devm_gpiod_get_index(dev, "sd_cd", 0);
 	if (IS_ERR(desc)) {
@@ -212,25 +206,15 @@ static int sdhci_acpi_add_own_cd(struct device *dev, struct mmc_host *mmc)
 		goto out;
 	}
 
-	err = gpiod_direction_input(desc);
-	if (err)
-		goto out_free;
-
-	irq = gpiod_to_irq(desc);
-	if (irq < 0) {
-		err = irq;
-		goto out_free;
-	}
+	gpio = desc_to_gpio(desc);
+	devm_gpiod_put(dev, desc);
 
-	flags = IRQF_TRIGGER_RISING | IRQF_TRIGGER_FALLING;
-	err = devm_request_irq(dev, irq, sdhci_acpi_sd_cd, flags, "sd_cd", mmc);
+	err = mmc_gpio_request_cd(mmc, gpio, 0);
 	if (err)
-		goto out_free;
+		goto out;
 
 	return 0;
 
-out_free:
-	devm_gpiod_put(dev, desc);
 out:
 	dev_warn(dev, "failed to setup card detect wake up\n");
 	return err;
-- 
1.8.3.2

